### PR TITLE
Article save switch

### PR DESCRIPTION
**Pull request by NathanielBennett**
This is a 0% a/b test to test "Save for later" functionality that we're going to test internally to start with that provides a 'save for later' link on content pages ( on desktop and wider ) 
. Once an article is saved, theres a link to indicate this. This link doesn't work yet, but is coming. But as this only being used internally, that should be OK

save-for-later

saved-articles

Logic: 

For signed in users, get the users current list of saved articles and check whether it contains the url of the current page. If it's isnt the 'Save article' link will  add it to the users saved articles via an ajax call to the identity api ( see below ), if it is display a 'saved articles' link. This is currently dead - but we're only using this internally for the moment. When a user is not signed in, the save article link takes them through the sign-in flow and saves the article, server-side to the identity api.

ID api notes

The identity api provides endpoints for retrieving and posting lists of articles to a synced prefs collection. Retrieving a users list of saved articles returns a json document in the form

``` json
{
    "version": "2015-04-01T00:09:06+11:00,
    "articles": [
        {
            "id": "http://m.thegulocal.com/world/2015/mar/31/death-sentence-killer-british-medical-students-borneo-zulkipli-abdullah",
            "shortUrl": "http://gu.com/p/475ajj",
            "date": "2015-04-01T00:09:06+11:00",
            "read": false
        }
    ]
}
```

where version is a timestamp. In order to save an article, we need to add another item to the articles array and post it back with the version value that was originally retrieved. This is because the id api uses "version" to manage saving by the same user across different devices. Ie, we need to do a get and a post. 

The caveat here is the case where a user has no previously saved documents ( syncedPrefs being a distinct collection in the identity api from users ). In this case, the api returns an error response with a status of 404 and a document like this

``` json
[{message:'Not found', description:'Resource not found'}]
```

So, in this case, we need to handle that specific response differently to other errors

There's probably some performance improvements to use local storage to limit id api calls by using local storage - but whilst we just play around with this internally, that should be OK

### CODE REVIEW

File: common/app/views/fragments/contentMeta.scala.html [hunk 1]
```diff
@@ -34,6 +34,8 @@
             <div class="u-h meta__number" data-discussion-id="@content.discussionId" data-commentcount-format="content" data-discussion-closed="@{!content.isCommentable}">
             </div>
         </div>
+        <div class="meta__save-for-later" data-link-name="meta-save-for-later" data-component=meta-save-for-later">
```
Comment (robertberry) [inline common/app/views/fragments/contentMeta.scala.html]: I think you're missing a `"` here

File: identity/app/controllers/SaveContentController.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.google.inject.Inject
+import common.ExecutionContexts
+import idapiclient.IdApiClient
+import play.api.mvc._
+import actions.AuthenticatedActions
+import services.{PlaySavedArticlesService, IdRequestParser, ReturnUrlVerifier}
+import utils.SafeLogging
+
+class SaveContentController @Inject() ( api: IdApiClient,
```
Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: What's this `Inject` shizzle?

File: static/src/javascripts/projects/common/modules/loyalty/save-for-later.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                    self.$saver.html('<a href="' + self.savedArticlesUrl + '" data-link-name="meta-save-for-later" data-component=meta-save-for-later">Saved Articles</a>');
+                } else {
+                    self.$saver.html('<a class="meta__save-for-later--link" data-link-name="meta-save-for-later" data-component=meta-save-for-later">Save for later</a>');
+                    bean.on(self.saveLinkHolder, 'click', '.meta__save-for-later--link', self.saveArticle.bind(self));
+                }
+            }
+        );
+    };
+
+    SaveForLater.prototype.hasŬserSavedArticle = function (articles, shortUrl) {
```
Comment (piuccio) [inline static/src/javascripts/projects/common/modules/loyalty/save-for-later.js]: Unicode FTW :wink:

File: static/src/javascripts/projects/common/modules/loyalty/save-for-later.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    };
+
+    SaveForLater.prototype.getSavedArticles = function () {
+        var self = this,
+            notFound  = [{message:'Not found', description:'Resource not found'}];
+
+        identity.getUsersSavedDocuments().then(
+            function success(resp) {
+                if (resp.status === 'error') {
+                    if (JSON.stringify(notFound) ===  JSON.stringify(resp.errors)) {
```
Comment (piuccio) [inline static/src/javascripts/projects/common/modules/loyalty/save-for-later.js]: I think it's better to compare `resp.errors[0].message === notFound.message && resp.errors[0].description === notFound.description`.
You can't guarantee that JSON.stringify maintains the same order of properties for two different objects.

File: static/src/javascripts/projects/common/modules/loyalty/save-for-later.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                        //Identity api needs a string in the format yyyy-mm-ddThh:mm:ss+hh:mm  otherwise it barfs
+                        var date = new Date().toISOString().replace(/\.[0-9]+Z/, '+00:00');
```
Comment (piuccio) [inline static/src/javascripts/projects/common/modules/loyalty/save-for-later.js]: Can we use `moment` for formatting dates?

File: static/src/javascripts/projects/common/modules/experiments/tests/save-for-later.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        this.audience = 0.0;
+        this.audienceOffset = 0;
+        this.successMeasure = '';
+        this.audienceCriteria = 'Interal only - we opt in';
+        this.dataLinkNames = '';
+        this.idealOutcome = '';
+
+        this.canRun = function () {
+            var isNotContent = /Network Front|Section/.test(config.page.contentType),
+                isDesktop = detect.getBreakpoint() === 'desktop' || detect.getBreakpoint() === 'wide';
```
Comment (phamann) [inline static/src/javascripts/projects/common/modules/experiments/tests/save-for-later.js]: Would be easier to use [`detect.isBrekpoint({min: ''desktop"})`](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/utils/detect.js#L305) here.

**Comment by gklopper** [comment]
:+1:

>This link doesn't work yet, but is coming.

I could write a link that does not work in a lot less code :laughing:

File: identity/app/controllers/SaveContentController.scala (hunk 2 above)
Comment (gklopper) [inline identity/app/controllers/SaveContentController.scala]: @daithiocrualaoich and I tried to convince those ID developers back in the day...

Comment (robertberry) [inline identity/app/controllers/SaveContentController.scala]: Convince them to what? Is this Guice or something horrid?

File: static/src/javascripts/projects/common/modules/loyalty/save-for-later.js (hunk 5 above)
Comment (NathanielBennett) [inline static/src/javascripts/projects/common/modules/loyalty/save-for-later.js]: @pluccio apparaently - even though the question keeps coming up - moments big and that's why we haven't got it. So no

Comment (piuccio) [inline static/src/javascripts/projects/common/modules/loyalty/save-for-later.js]: @NathanielBennett Interesting, I saw `moment` in `package.json` and I thought we were using it already. But it's not in `requirejs.js` and I can't find any place where we actually use it. Probably a forgotten dependency. Fine

**Pull request merged**